pkg/ebpf/ingress_node_firewall_loader.go

```diff
@@ -55,43 +59,295 @@ func NewIngNodeFwController() (*IngNodeFwController, error) {
 
-	if err := infc.makeIngressFwRulesMap(ingFireWallConfig, isDelete, ifId); err != nil {
-		return fmt.Errorf("Failed to create map firewall rules: %v on if %d", err, ifId)
+	// Add/update all keys. This is an idempotent action and non-existing keys are added whereas existing keys
+	// are updated.
+	if err := infc.addOrUpdateRules(ebpfKeyToRules); err != nil {
+		return err
 	}
 
+	// Generate ingress node few events.
 	infc.ingressNodeFwEvents()
 
 	return nil
 }
 
+// addOrUpdateRules is a small method containing this limited set of functionality to facilitate unit testing of
+// this part of code.
+func (infc *IngNodeFwController) addOrUpdateRules(ebpfKeyToRules map[BpfLpmIpKeySt]BpfRulesValSt) error {
+	for ebpfKey, ebpfRules := range ebpfKeyToRules {
+		log.Printf("Adding or updating ingress firewall rules for key %v", ebpfKey)
+		if err := infc.objs.BpfMaps.IngressNodeFirewallTableMap.Update(ebpfKey, ebpfRules, ebpf.UpdateAny); err != nil {
+			return fmt.Errorf("Failed Adding/Updating ingress firewall rules: %v", err)
+		}
+	}
+	return nil
+}
+
+// GetStatisticsMap returns the statistics map of the object.
 func (infc *IngNodeFwController) GetStatisticsMap() *ebpf.Map {
 	return infc.objs.IngressNodeFirewallStatisticsMap
 }
 
-// makeIngressFwRulesMap convert IngressNodeFirewallRules into eBPF format which matched what the
-// kerenl hook will be using.
-func (infc *IngNodeFwController) makeIngressFwRulesMap(ingFirewallConfig ingressnodefwiov1alpha1.IngressNodeFirewallRules,
-	isDelete bool, ifId uint32) error {
+// IngressNodeFwAttach attaches the eBPF program to a given list of interfaces and pins them to different pinDirs.
+// For each provided interface name:
+// i)   Look up the network interface by name.
+// ii)  Attach the program to the interface.
+// iii) Pin the XDP program.
+func (infc *IngNodeFwController) IngressNodeFwAttach(ifacesName ...string) error {
 	objs := infc.objs
-	rules := BpfRulesValSt{}
+	for _, ifaceName := range ifacesName {
+		if _, ok := infc.links[ifaceName]; ok {
+			klog.Infof("Interface %s is already attached and managed, skipping", ifaceName)
+			return nil
+		}
+		// Look up the network interface by name.
+		iface, err := net.InterfaceByName(ifaceName)
+		if err != nil {
+			return fmt.Errorf("lookup network iface %q: %s", ifaceName, err)
+		}
+
+		// Attach the program.
+		l, err := link.AttachXDP(link.XDPOptions{
+			Program:   objs.IngressNodeFirewallProcess,
+			Interface: iface.Index,
+		})
+		if err != nil {
+			if strings.Contains(err.Error(), xdpEBUSYErr) {
+				return fmt.Errorf("Interface %s is already attached", ifaceName)
+			}
+			return fmt.Errorf("could not attach XDP program: %s", err)
+		}
+		// Pin the XDP program.
+		lPinDir := path.Join(infc.pinPath, ifaceName+"_link")
+		if err := l.Pin(lPinDir); err != nil {
+			return fmt.Errorf("failed to pin link to pinDir %s: %s", lPinDir, err)
```

Comment:
we need to append error to aggr errors slice and continue to the next interface and same for the above errors ?

Reply:
done